skimage/filters/__init__.py

```diff
@@ -1,74 +1,31 @@
-from .lpi_filter import inverse, wiener, LPIFilter2D
-from ._gaussian import (gaussian, _guess_spatial_dimensions,
-                        difference_of_gaussians)
-from .edges import (sobel, sobel_h, sobel_v,
-                    scharr, scharr_h, scharr_v,
-                    prewitt, prewitt_h, prewitt_v,
-                    roberts, roberts_pos_diag, roberts_neg_diag,
-                    laplace,
-                    farid, farid_h, farid_v)
-from ._rank_order import rank_order
-from ._gabor import gabor_kernel, gabor
-from .thresholding import (threshold_local, threshold_otsu, threshold_yen,
-                           threshold_isodata, threshold_li, threshold_minimum,
-                           threshold_mean, threshold_triangle,
-                           threshold_niblack, threshold_sauvola,
-                           threshold_multiotsu, try_all_threshold,
-                           apply_hysteresis_threshold)
-from .ridges import (meijering, sato, frangi, hessian)
-from . import rank
-from ._median import median
-from ._sparse import correlate_sparse
-from ._unsharp_mask import unsharp_mask
-from ._window import window
-from ._fft_based import butterworth
+from .._shared import lazy
 
-
-__all__ = ['inverse',
-           'correlate_sparse',
-           'wiener',
-           'LPIFilter2D',
-           'gaussian',
-           'butterworth',
-           'difference_of_gaussians',
-           'median',
-           'sobel',
-           'sobel_h',
-           'sobel_v',
-           'scharr',
-           'scharr_h',
-           'scharr_v',
-           'prewitt',
-           'prewitt_h',
-           'prewitt_v',
-           'roberts',
-           'roberts_pos_diag',
-           'roberts_neg_diag',
-           'laplace',
-           'farid',
-           'farid_h',
-           'farid_v',
-           'rank_order',
-           'gabor_kernel',
-           'gabor',
-           'try_all_threshold',
-           'meijering',
-           'sato',
-           'frangi',
-           'hessian',
-           'threshold_otsu',
-           'threshold_yen',
-           'threshold_isodata',
-           'threshold_li',
-           'threshold_local',
-           'threshold_minimum',
-           'threshold_mean',
-           'threshold_niblack',
-           'threshold_sauvola',
-           'threshold_triangle',
-           'threshold_multiotsu',
-           'apply_hysteresis_threshold',
-           'rank',
-           'unsharp_mask',
-           'window',
-           ]
+__getattr__, __dir__, __all__ = lazy.attach(
+    __name__,
+    submodules={'rank'},
+    submod_attrs={
+        'lpi_filter': ['inverse', 'wiener', 'LPIFilter2D'],
+        '_gaussian': ['gaussian', '_guess_spatial_dimensions',
+                      'difference_of_gaussians'],
+        'edges': ['sobel', 'sobel_h', 'sobel_v',
```

Comment:
My general worry about things like this, is that it breaks people's internal knowledge of what python is according to other tutorials.

The barrier to contribution is thus increased from:
* Someone that knows how to contribute a good algorithm
to
* Someone who knows all the funny tricks used at scikit-image to get marginal gains.

Currently the import time is dominated by two packages:

![image](https://user-images.githubusercontent.com/90008/124521832-f3476800-ddbe-11eb-81b1-81e126d6ec25.png)

```
# You might need to redirect 1 not 2, but for some reason I couldn't redirect both...
python -X importtime -c "import skimage" 2>import.log
tuna import.log
```

Reply:
From that graph, we are limited by importing numpy, and requests from pooch. 

I think our efforts would be better spent making pooch more "on demand" and biting the bullet on numpy. Numpy seems pretty standard so far as all these libraries go.

Reply:
Leaving it here for further exploration, but this PR is pretty specific about a submodule that is pretty slow to load.

I think a fair benchmark would be to compare the import time of a few key functions:

```
python -X importtime -c "import numpy; import pooch; from skimage.filters.edges import sobel" 2>import.log && tuna import.log
```
Should be able to help you compare the import time of the specific edge filter, when all other functions are "lazy"

Reply:
@hmaarrfk although I totally get your concerns, I think the benefits are bigger than you're suggesting, because now people are able to `import skimage; result = skimage.filters.gaussian(image)`. I think that actually *helps* a lot of beginners who are not aware that submodules don't get imported automatically.

Reply:
that is an important benefit